File: opcua-webservice-spring/core/src/main/java/org/opcfoundation/webserver/service/transactions/base/UaTransactionManager2.java
package org.opcfoundation.webserver.service.transactions.base;

import org.eclipse.milo.opcua.stack.core.StatusCodes;
import org.eclipse.milo.opcua.stack.core.UaRuntimeException;
import org.eclipse.milo.opcua.stack.core.types.builtin.NodeId;

import java.util.*;
import java.util.concurrent.CompletableFuture;

public class UaTransactionManager2<T_Request,T_Response> {
    private final List<UaTransaction2<T_Request, T_Response>> transactions;
    private int currentTransactionIndex;

    public UaTransactionManager2()
    {
        transactions = new ArrayList<>();
        currentTransactionIndex = 0;
    }

    public void addTransaction(UaTransaction2<T_Request, T_Response> transaction)
    {
        transactions.add(transaction);
    }

    public List<T_Response> getMergedResults() throws UaRuntimeException
    {
        Map<Integer,T_Response> results = new TreeMap<>();

        for (UaTransaction2<T_Request,T_Response> item : transactions)
        {
            if (item.getResults().size() != item.getHandleIds().size()) {
                throw new UaRuntimeException(StatusCodes.Bad_InternalError);
            }

            for (int i=0;i<item.getHandleIds().size(); ++i)
            {
                results.put(item.getHandleIds().get(i), item.getResults().get(i));
            }
        }

        return new ArrayList<>(results.values());
    }

    public CompletableFuture<UaTransactionManager2<T_Request,T_Response>> execute()
    {
        if (transactions.size() <= currentTransactionIndex)
        {
            return CompletableFuture.completedFuture(this);
        }

        return transactions.get(currentTransactionIndex).
                execute().
                thenCompose(V -> executeComplete());
    }

    public static List<Integer> getNsIndexes(List<NodeId> nodeIds)
    {
        Set<Integer> nsIndexes = new HashSet<>();
        for (NodeId item : nodeIds)
        {
            nsIndexes.add(item.getNamespaceIndex().intValue());
        }

        return new ArrayList<>(nsIndexes);
    }

    public static List<Integer> getHandleIds(List<NodeId> nodeIds, Integer nsIndex)
    {
        ArrayList<Integer> ret = new ArrayList<>();

        for (int i=0; i<nodeIds.size(); ++i)
        {
            if (nodeIds.get(i).getNamespaceIndex().intValue() == nsIndex) ret.add(i);
        }

        return ret;
    }

    private CompletableFuture<UaTransactionManager2<T_Request,T_Response>> executeComplete()
    {
        currentTransactionIndex++;
        return execute();
    }
}

File: opcua-webservice-node/src/server/service/transactions/base/UaTransactionManager2.ts
import { StatusCodes } from "opcua-webapi";
import { makeUaStatusCode, UaError, UaNodeId } from "opcua-webapi-ts";
import { UaTransaction2 } from "./UaTransaction2";

export class UaTransactionManager2<TRequest, TResponse> {
    private readonly transactions: Array<UaTransaction2<TRequest, TResponse>>;

    constructor() {
        this.transactions = [];
    }

    addTransaction(transaction: UaTransaction2<TRequest, TResponse>): void {
        this.transactions.push(transaction);
    }

    getMergedResults(): Array<TResponse> {
        const results = new Map<number, TResponse>();
        const mergedResults: Array<TResponse> = [];

        for (const item of this.transactions) {
            if (item.getResults().length !== item.handleIds.length) throw new UaError(makeUaStatusCode(StatusCodes.BadUnexpectedError));

            for (let index = 0; index < item.handleIds.length; index++) {
                results.set(item.handleIds[index], item.getResults()[index]);
            }
        }

        for (let index = 0; index < results.size; index++) {
            if (!results.has(index)) throw new UaError(makeUaStatusCode(StatusCodes.BadUnexpectedError));
            mergedResults.push(results.get(index) as TResponse);
        }

        return mergedResults;
    }

    async execute(): Promise<void> {
        for (let transaction of this.transactions) {
            await transaction.execute();
        }
    }

    static getNsIndexes(nodeIds: Array<UaNodeId>): Array<number> {
        const nsIndexes = new Set<number>();

        for (const item of nodeIds) {
            nsIndexes.add(item.nsIndex);
        }

        return Array.from(nsIndexes);
    }

    static getHandleIds(nodeIds: Array<UaNodeId>, nsIndex: number): Array<number> {
        const ret: Array<number> = [];

        for (let index = 0; index < nodeIds.length; index++) {
            if (nodeIds[index].nsIndex === nsIndex) ret.push(index);
        }

        return ret;
    }
}